[5기] 3주차 Wordle 과제 제출 - 호아

hoa0217 opened the pull request:
## 잼아팀🔥 페어프로그래밍
- 페어 : @jongmin4943 와잼
- 방식 : 10분 타이머를 이용, 드라이버와 네비게이터 교체, 40분 or 60분마다 10분 휴식
- 진행 :
  - 6/8(토) 13:30 ~ 18:30 (5시간) - 오프라인
  - 6/9(일) 14:00 ~ 16:00 (2시간) - 오프라인
  - 6/12(수) 20:05 ~ 22:45 (2시간40분) - 온라인
  - 6/13(목) 20:00 ~ 22:00 (2시간) - 온라인
  - 6/15(토) 14:00 ~ 17:30 (3시간30분) - 오프라인
  - 6/17(월) 20:20 ~ 22:20 (2시간) - 온라인

## 배운점과 잘한점
넥텝 리뷰어이신 와잼님의 위대함을 느낄 수 있었습니다.. 특히 와잼님의 객체지향에 대한 철학과 코드 스타일은 정말 따라하고 싶을 정도로 배울점이 많았습니다.

프로덕션 코드를 짜기전, 테스트 코드를 먼저 짜는 TDD 방식으로 진행하였는데 확실히 테스트 코드가 탄탄하니, 부담없이 테스트를 돌려보며 리팩토링을 진행할 수 있었습니다. 테스트코드 짱짱 😎

저희팀은 [객체지향 생활 체조 원칙 9가지](https://jamie95.tistory.com/99)를 빡빡하게 지켜보자는 약속을 갖고 페어를 진행했으며 동시에 getter를 최대한 지양해보려고 했는데 이 과정에서 더욱더 객체끼리 협력할 수 있는 코드가 나온것 같습니다!!

> 중복되는 부분은 템플릿 메서드패턴으로 중복을 제거하는 와잼님의 꾸울팁에 정말 박수를 쳤고 앞으로도 유용하게 사용할것 같습니다! 👍

## 아쉬운점
서로 바쁜 시기라 더 많이 페어를 진행하니 못한 것이 아쉬웠습니다.. ㅠㅠ 오늘도 카페마감시간때문에.. 흑흑.. 파라미터 3개를 받는 부분을 해결했다면 더 좋았을거같네요.. 

제가 한 방향에 꽂히면 그게 옳고 그른지 판단력이 흐려지는데 해당 과정에서 저의 커뮤니케이션 방식(?)을 반성했습니다..!! (소프트스킬 책 읽어야겠네요.) 그래도 최대한 많은 의견을 수용해주신 와잼님께 감사를 🥹

시간을 고정하고자 `staticMock`을 사용하였는데 요부분이 Gradle의 의존성을 추가해야하더라구요. 이미 추가한 상태에서 외부라이브러리를 사용하면 안되는 것을 알았던게 아쉬웠습니다.. 그래도 덕분에 실제 시간에 의존적이지 않은 테스트케이스를 작성할 수 있었습니다.

## 많은 시간을 투자한 부분
첫날 페어를 진행하면서 용어사전을 정의하고 클래스다이어그램을 그리며 설계를 하면서 많은 시간을 쏟았습니다. 보다 복잡한 케이스는 `ll현상`이라는 저희끼리만의 언어도 도출하기도 하였으며, 확실히 토론을 진행하며 의견을 나누니 서로 가지고있는 이해도 수준을 맞출 수 있었습니다.

그 후 코드를 작성하면서는 가장 복잡했던 `ll현상`.. 처음엔 Word를 비교하는 과정을 절차지향적으로 알고리즘 풀듯이 한번 구현한 후 나중에 리팩토링하는 형식으로 진행하였는데 해당 방식을 어떻게 하면 깔끔하게 풀어낼 수 있을지 고민하는데 시간을 많이 썼습니다. (미리 생각해왔어야했는데 뇌가 알고리즘 푸는데 적응이 되어서 그런지 잘 안되더라구요..)

## 페어에게 하고 싶은 말
와잼님께 정말 질문을 많이 했던 것 같은데, 그럴 때마다 논리적으로 잘 설명해주셔서 너무너무  감사했습니다..!! 많은 인사이트를 얻어갈 수 있었어요. 항상 페어하는 시간이 기대되었고 끝나는 시간이 아쉬웠습니다 ㅜㅜ 다음번에 또 페어할 수 있는 기회가 있었으면 좋겠네요!! 일취월장하겠습니다!!! 💪🏻 넘넘 고생하셨어요 🙂

jongmin4943 committed:
test: 통합 테스트 - 성공케이스 작성

hoa0217 committed:
docs(README): 게임진행순서, 용어사전 정리

hoa0217 committed:
docs(README): 모델링(클래스 다이어그램) 작성

hoa0217 committed:
feat(Position): Position 클래스 생성

- 위치에 대한 책임을 가진 클래스

hoa0217 committed:
feat(Letter): Letter 클래스 생성

- 알파벳과 위치를 가진 클래스

hoa0217 committed:
feat(Word): Word 클래스 생성

- List<Letter> 를 가지는 클래스

hoa0217 committed:
feat(Result): Result 클래스 생성

- Tile, Position 을 가지는 클래스

hoa0217 committed:
refactor(Letter): isSameAlphabetDifferentPosition 메서드명 변경

- isSameAlphabet로 간결하게 변경
- 해당 메서드가 너무 많은 컨텍스트를 알고있음

hoa0217 committed:
fix(Result): equals hashCode 구현

hoa0217 committed:
feat(Result): Results 클래스 생성

- Result 일급컬렉션 클래스

hoa0217 committed:
test(Result): Result Fixture 분리

hoa0217 committed:
feat(Word): compare 메서드 추가

hoa0217 committed:
fix(Word): compare 메서드에 노랑타일로직 추가

hoa0217 committed:
test(Word): 글자와 위치가 일부가 같은 경우 테스트 추가

hoa0217 committed:
test(Word): 글자와 위치가 전부 다른 경우 테스트 추가

hoa0217 committed:
fix(Results): 필드 배열에서 List 로 변경

hoa0217 committed:
refactor(Results): 필드 구현체 SortedSet 으로 변경

hoa0217 committed:
fix(Word): 같은 문자가 2개 입력되었을 때, 해당 문자가 정답에 하나만 존재하지만 위치가 틀린 경우 로직 추가

hoa0217 committed:
test(Time): LocalDate 모킹 유틸 추가

hoa0217 committed:
feat(AnswerFormula): AnswerFormula calculate 구현

hoa0217 committed:
feat(FileReader): FileReader 구현

- resource 밑에 있는 파일을 읽어온다

hoa0217 committed:
feat(FileReader): FileReader 구현

- resource 밑에 있는 파일을 읽어온다

hoa0217 committed:
fix(AnswerFormula): calculate 메서드 반환 타입 int로 변경

- index로 사용되는 용도이므로 변경

hoa0217 committed:
fix(AnswerFormula): calculate 메서드 파라미터 예외처리 추가

hoa0217 committed:
feat(WordBook): 파일 기반 WordBook 클래스 생성

- 단어장 책임을 가진 클래스

hoa0217 committed:
feat(Alphabet): Alphabet 클래스 생성

- 알파벳을 소문자로 저장하는 클래스

hoa0217 committed:
refactor(Letter): Alphabet을 값객체(VO)로 갖는다.

hoa0217 committed:
Merge remote-tracking branch 'origin/jamwa' into jamwa

# Conflicts:
#	src/main/java/wordle/infra/FileReader.java
#	src/test/java/wordle/infra/FileReaderTest.java

hoa0217 committed:
feat(Record): 단어 비교 기록 클래스 생성

hoa0217 committed:
test(Result): 초록색 타일 여부 확인 테스트 추가

hoa0217 committed:
fix(Result): 초록색,회색 타일 여부 추가

hoa0217 committed:
feat(InputView): 사용자 입력을 받는 클래스 생성

hoa0217 committed:
feat(OutputView): 출력하는 클래스 생성

hoa0217 committed:
feat(WordBook): 단어를 찾는 메서드 추가

hoa0217 committed:
test(WordBook): 시간관련 의존성 제거

hoa0217 committed:
fix(Position): 음수 예외처리 추가

hoa0217 committed:
fix(Alphabet): 알파벳은 대소문자를 구분하지 않는다

hoa0217 committed:
fix(ConsoleOutputView): 성공, 실패 종료문구 분리

hoa0217 committed:
fix(Record): existAllGreen, isCountOver 메서드 추출

hoa0217 committed:
feat(Wordle): 게임완성

hoa0217 committed:
fix(Exception): 예외 정리

- 게임 비정상적 종료 처리 추가

hoa0217 committed:
feat(FileReader): FileReader 구현

- resource 밑에 있는 파일을 읽어온다

hoa0217 committed:
refactor(Word): 코드 정리

- 안쓰는 메서드 정리 및 포맷 맞추기

hoa0217 committed:
test: 테스트에 사용되는 리소스(word.text)및 케이스 변경

hoa0217 committed:
refactor: 코드 정리

- final 변수 생성 및 메서드 정리

hoa0217 committed:
test: 콘솔 통합테스트 시간 고정

hoa0217 committed:
docs: 리드미 정리

- 용어 사전 수정 및 클래스 다이어그램 수정

woo-yu commented on build.gradle:
```diff
@@ -16,6 +16,7 @@ repositories {
 }
 
 dependencies {
+    testImplementation 'org.mockito:mockito-inline:3.12.0'
```
테스트하실때 mock을 활용하셨군요! ㅇ_ㅇ

hoa0217 replied:
네 맞습니다! staticMock을 사용하려고 의존성을 추가했습니다~~

woo-yu commented on docs/README.md:
```diff
@@ -0,0 +1,163 @@
+- 6번 안에 맞추면 게임을 종료한다.
+- 6번 안에 맞추지 못하면 그래도 종료한다.
+
+## 용어 사전
```
크 역시 ddd 영재 호아님 👍 정리하신거 멋집니다!

hoa0217 replied:
감사합니당!!

woo-yu commented on docs/README.md:
```diff
@@ -0,0 +1,163 @@
+- 입력받은 문자와 정답을 비교한다.
+- 비교 결과는 타일이 초록색/노란색/회색 중 하나로 바뀌면서 표현된다.
+    - 맞는 글자는 초록색, 위치가 틀리면 노란색, 없으면 회색
+    - 같은 문자가 2개 입력되었을 때, 해당 문자가 정답에 하나만 존재하지만 위치가 틀린 경우 첫번 째 문자만 노란색으로 표시된다.
```
같은 문자가 3개 입력되었을 때, 해당 문자가 정답에 하나만 존재하면 어떻게 되나요? 😄

hoa0217 replied:
같은 문자가 n개 입력되었을 때, 하나만 존재하면 첫번 째 문자만 노란색으로 표시됩니다! 문서 수정할게요~

woo-yu commented on docs/README.md:
```diff
@@ -0,0 +1,163 @@
+|--------|----------------|------------------------------------------------|
+| 워들     | Wordle         | 5글자 영어 단어 맞추기 게임                               |
+| 단어장    | Word Book      | 이 게임에서 사용될 수 있는 단어 모음                          |
+| 입력 단어  | Input Word     | 플레이어가 입력하는 5글자 단어                              |
+| 정답 단어  | Answer Word    | 오늘 게임의 정답인 5글자 단어                              |
```
클래스명중에 Word가 있는데 이 Word는 입력단어와 정답 단어를 포괄하는 개념인가요? 😄

hoa0217 replied:
넵 초반에는 InputWord와 AnswerWord 클래스를 분리하려고 했으나, 굳이 그럴 필요가 없겠더라구요~ 개발하면서 모델링이 계속 변경되었는데 지속적 업데이트의 중요성을 느낍니다 ㅠㅠ

woo-yu commented on docs/README.md:
```diff
@@ -0,0 +1,163 @@
+
+## 모델링
+
+### 클래스 다이어그램
```
클래스 다이어그램까지 👍👍👍👍👍

woo-yu commented on src/main/java/wordle/domain/Record.java:
```diff
@@ -0,0 +1,45 @@
+public class Record implements Iterable<Results> {
+
+    public static final int MAX_COUNT = 6;
+    private final List<Results> record;
```
아닛 일급컬렉션 구현까지 👍👍👍👍👍

woo-yu commented on src/main/java/wordle/application/Wordle.java:
```diff
@@ -0,0 +1,72 @@
+            return;
+        }
+        outputView.failEnd(record);
+    }
```
record에는 종료되었는지 여부를 isEnd() 조건을 잘 만들어주셨는데요 👍
정답을 맞췄을때 종료와 실패했을때 종료의 분기처리가 필요해지는군요!
혹시 record에서 `existAllGreen`라는 rule을 물어보기보다는 정말로 정답을 맞췄는지 물어보는건 어떨까요? 😄

woo-yu commented on src/main/java/wordle/domain/Record.java:
```diff
@@ -0,0 +1,45 @@
+        return record.size() >= MAX_COUNT;
+    }
+
+    public boolean existAllGreen() {
```
`existAllGreen` 요 말은 워들 고수인 우리들은 😎 보자마자 알겠군요. 
- all green? 이라면 정답 한줄을 맞췄나보군. 그러면 정답을 맞췄는지 여부를 확인하는구나 라고 추측할 수 있겠네요!

워들의 룰을 잘 모르는 사람이 만약에 정답을 맞췄는지 여부를 찾으려면 요 문구를 보고 바로 알 수 있을까요? 🤔
조금 더 직관적으로 '정답을 맞췄는지'라는 네이밍으로 물어볼 수 있을까요?

hoa0217 replied:
rule로 네이밍을 하기보다는 정답을 맞췄니를 나타낼 수 있는 existAnswer가 좋겠네요!

woo-yu commented on src/main/java/wordle/domain/Record.java:
```diff
@@ -0,0 +1,45 @@
+        record.add(results);
+    }
+
+    public boolean isEnd() {
+        return existAllGreen() || isCountOver();
+    }
```
요기에 종료조건 두 가지를 알차게 나눠주셨군요 😄
둘의 차이점은 뭘까요? 위에서 언급했던 '정답을 맞췄을때 종료와 실패했을때 종료의 분기처리'를 요기서도 잘 해낼 수 있겠군요 😄

hoa0217 replied:
우유님 코멘트보고 계속 고민해보았는데.. isEnd라는 조건으로 게임 run을 돌릴게 아닌 isCountOver이라는 조건으로 돌리면 분기처리를 충분히 해낼 수 있겠더라구요!! 코드에 반영했습니당 🤩

hoa0217 replied:
또한 기억을 더듬어 보니.. 게임 종료 조건을 `isEnd`로 한번에 묶어버렸던 이유는 while문에서 아래처럼 들여쓰기가 두단계 이상이 되지않으려고 while조건문에 묶어주었던건데

어플리케이션단 코드는 어차피 절차지향적일 수 밖에 없고, 오히려 아래와 같이 수정하는게 게임을 이해하는데 더 쉬울거같아서 수정하였습니다!!

```java
private boolean runGame(Word answerWord) {
        while (!record.isCountOver()) {
            outputView.showRecord(record);
            handleWrongAnswer(() -> processRound(answerWord));

            if (record.existAnswer()) {
                return true;
            }
        }
        return false;
    }
```

woo-yu commented on src/main/java/wordle/application/Wordle.java:
```diff
@@ -0,0 +1,72 @@
+            outputView.wrongAnswer();
+        } catch (WordleException e) {
+            outputView.unexpectedEnd();
+        }
```
크 exception과 각각의 화면출력요소를 남기는 센스 👍👍
outputView에서 wrongAnswer와 unexpectedEnd는 서로 출력 메시지만 다르더라고요!
exception에서 message를 담아서 출력해주는 방식은 혹시 어떻게 생각하시나요 😄

hoa0217 replied:
아예 예외메세지를 담아서 outputView에 넘겨주는 방법도 있겠네요~

woo-yu commented on src/main/java/wordle/domain/FileWordBook.java:
```diff
@@ -0,0 +1,38 @@
+        return words.contains(word);
+    }
+
+    @Override
+    public Word find(String target) {
+        Word targetWord = new Word(target);
+        return words.stream()
+                .filter(targetWord::equals)
+                .findFirst()
+                .orElseThrow(WordNotExistException::new);
+    }
```
오... 단어가 없을 때 대한 예외처리도 WordBook이 직접 하는군요? ㅋㅋㅋㅋ
FileWordBook이 인프라적인 요소라고 생각해보면, 예를들어 db에서 조회해오는 repository라고 생각해보면.... repo에서 직접 예외를 던지시는 편이신가욧? 😄

hoa0217 replied:
헉...... WordBook이 레포라는 관점으로 보면 어플리케이션 서비스에서 던지는게 맞겠네요.. 역시 👍

woo-yu commented on src/main/java/wordle/domain/Letter.java:
```diff
@@ -0,0 +1,40 @@
+        return letter.alphabet.equals(this.alphabet);
+    }
+
+    public Position getPosition() {
+        return position;
+    }
```
오 위치를 담고있군요 👍👍

woo-yu commented on src/main/java/wordle/domain/Alphabet.java:
```diff
@@ -0,0 +1,37 @@
+
+    public Alphabet(char alphabet) {
+        char lowerAlphabet = Character.toLowerCase(alphabet);
+        if (lowerAlphabet < MIN_ALPHABET || lowerAlphabet > MAX_ALPHABET) {
```
알파벳에 min과 max라는 표현을 사용할수도 있군요! 뭔가 숫자 표기만 보다가 문자표기는 신선하네요 👍

woo-yu commented on src/main/java/wordle/exception/InvalidPositionException.java:
```diff
@@ -0,0 +1,5 @@
+package wordle.exception;
+
+public class InvalidPositionException extends WordleInvalidInputException {
```
WordleInvalidInput은 사용자 입력에 대한 예외처리로 이해하고 있었는데욧..!!
InvalidPosition은 사용자가 무슨 잘못(?)을 해야 발생하나요? 😄

hoa0217 replied:
사용자가 잘못입력했을 경우보다는, 해당 클래스를 개발자가 잘못 생성했을 때(?) 발생하겠네요!!

woo-yu commented on src/main/java/wordle/application/Wordle.java:
```diff
@@ -0,0 +1,72 @@
+        }
+    }
+
+    private void processTurn(Word answerWord) {
```
각 단계의 명칭이 Turn이군요! 😄
Results의 용어사전에는 '라운드가 진행될 때 마다 누적된 결과모음'라고 적혀있는데 요기서 말하는 라운드와 턴은 동일한 단어일까요?

hoa0217 replied:
ㅎㅎㅎㅎㅎ 메서드명을 processRound로 변경해야겠네요

woo-yu commented on src/main/java/wordle/domain/Results.java:
```diff
@@ -0,0 +1,32 @@
+
+public class Results implements Iterable<Result> {
+
+    private final SortedSet<Result> results;
```
오..!! 뭔가 iterable한 collection은 list가 먼저 생각난거같은데 set을 사용하게 되신 이유가 궁금합니다 😄
중복제거가 필요한 요소들이 있을까요?

hoa0217 replied:
TreeSet을 사용하려고 해당 인터페이스를 사용하였습니다! Position 순서대로 자동 정렬하려구요~

woo-yu commented on src/main/java/wordle/domain/Results.java:
```diff
@@ -0,0 +1,32 @@
+                .anyMatch(result -> result.isSamePosition(position));
+    }
+
+    public boolean isAllGreen() {
```
요기에도 'allGreen'이라는 용어가 사용되는군요!

hoa0217 replied:
isAnswer이 더 좋은 메서드명이겠네요 ㅎㅎ

woo-yu commented on src/main/java/wordle/domain/Result.java:
```diff
@@ -0,0 +1,56 @@
+
+import java.util.Objects;
+
+public class Result implements Comparable<Result> {
```
크 Comparable센스 👍

woo-yu commented on src/main/java/wordle/domain/Results.java:
```diff
@@ -0,0 +1,32 @@
+        results.add(result);
+    }
+
+    public boolean isCheckedPosition(Position position) {
+        return results.stream()
+                .anyMatch(result -> result.isSamePosition(position));
+    }
```
요게 왜 들어갔을까 한참 고민했는데 Word.process()가 여러번 동작할 수 있어서 위치를 선점하지말라고 그랬군요!
Results가 지닌 results의 index로 생각하는게 아니라 각 result마다의 position으로 검토하는군요 👍👍👍

woo-yu commented on src/main/java/wordle/domain/Word.java:
```diff
@@ -0,0 +1,105 @@
+        }
+    }
+
+    public Results compare(Word inputWord) {
+        WordComparator wordComparator = new WordComparator(this.letters);
+        return wordComparator.compare(inputWord);
+    }
```
요기 부분에서 조금 헷갈렸던건 Input Word와 Answer Word의 역할 차이였던거 같아요!
용어사전에서는 구분되어있던 단어다보니 코드상에서도 둘이 구분되어있다고 생각했는데, Word 하나로 통합되어있다보니 😄

Answer과 Input을 비교한다는 느낌보다는 Word 내에서 inputWord와 비교할 수 있다..?!
input-input도 비교할 수 있나 생각이 들더라고요. ㅋㅋㅋㅋㅋ

hoa0217 replied:
맞아요 ㅠㅠ 클래스를 나누지 않았다보니.. 명시적으로 구분할 수 있는 방법을 찾아봐야겠네요!

woo-yu commented on src/main/java/wordle/domain/Word.java:
```diff
@@ -0,0 +1,105 @@
+        private final Results results;
+
+        public WordComparator(List<Letter> letters) {
+            this.pendingLetters = new ArrayList<>(letters);
```
pendingLetters라는 이름 좋네요 👍

woo-yu commented on src/main/java/wordle/domain/Word.java:
```diff
@@ -0,0 +1,105 @@
+            return results;
+        }
+
+        private void process(Letter targetLetter, Predicate<Letter> predicate, Tile tile) {
```
크 👍👍 predicate라는 전략 사용

woo-yu commented on src/main/java/wordle/domain/AnswerFormula.java:
```diff
@@ -0,0 +1,7 @@
+package wordle.domain;
+
+@FunctionalInterface
```
오 functional interace까지 👍

woo-yu commented on src/main/java/wordle/infra/FileReader.java:
```diff
@@ -0,0 +1,33 @@
+
+    public List<String> readByLine(String filePath) {
+        if (filePath == null || filePath.isBlank()) {
+            throw new FileReadFailException();
```
파일 읽으려면 넘어야할 산이 많군요? ㅋㅋㅋㅋㅋㅋ
뭔가 예외가 발생했을때 어디서 잘못되었을까?를 확인해볼때 주로 어느 메시지인지를 많이 확인하는거 같아요.
요기에도 각각이 왜 실패했는지 메시지를 담을 수 있을까요? 😄

hoa0217 replied:
맞아요 예외에 메세지를 대부분 빼먹었네요 ㅎㅎ 수정하겠습니다.

woo-yu commented on src/main/java/wordle/ui/ConsoleInputView.java:
```diff
@@ -0,0 +1,17 @@
+
+import java.util.Scanner;
+
+public class ConsoleInputView implements InputView {
```
Console 네이밍 너무 좋아요 👍

woo-yu commented on src/test/java/wordle/TimeTestSupporter.java:
```diff
@@ -0,0 +1,22 @@
+import org.mockito.MockedStatic;
+import org.mockito.Mockito;
+
+public class TimeTestSupporter {
```
시간때문에 Mock이 사용된거였군요! 늘 우리를 괴롭하는 .now() 😄
'테스트하기 좋은 코드'를 고민할때 요런 제어할 수 없는 static 요소, 그중에서도 시간때문에 항상 고민이 많습니다.
저는 이번 wordle을 만들땐 호아님의 Wordle과 같은 역할을 하는 객체를 만들때 오늘날짜를 집어넣었어요. ㅋㅋㅋㅋㅋㅋ 
게임도 오늘실행한 게임과 내일 실행한 게임이 다르겠지! 라고 생각해서요. 🥹

woo-yu commented on src/test/java/wordle/domain/AlphabetTest.java:
```diff
@@ -0,0 +1,33 @@
+public class AlphabetTest {
+
+    @Test
+    void 알파벳은_대소문자를_구분하지_않는다() {
```
역시 테스트는 한글로 짜야 제맛이죠 😄

woo-yu commented on docs/README.md:
```diff
@@ -0,0 +1,163 @@
+    - 알파벳이 아닌 경우 재입력을 받는다.
+- 입력받은 문자와 정답을 비교한다.
+- 비교 결과는 타일이 초록색/노란색/회색 중 하나로 바뀌면서 표현된다.
+    - 맞는 글자는 초록색, 위치가 틀리면 노란색, 없으면 회색
```
리뷰를 다하고 돌아보니... 여기에 맞는글자는 초록색이라고 아예 설명을 정해두고 가셨던거였군요! ㅋㅋㅋㅋㅋ
컬러가 바뀌면 allBlue 요런 네이밍으로 전체이동...되는건가요 😉

hoa0217 replied:
아이쿠 ㅎㅎㅎㅎ 뷰에 대한 책임을 대부분 콘솔에 몰아 놓고 메서드명에 박아버렸네요 수정해야겠어요!!

woo-yu reviewed (approved):
역시 ddd 잘알(?) 호아님! 코드 보면서 많이 배우고갑니다 🙇‍♀️
(리뷰하면서 제것도 슬쩍 고치게 되네요. ㅋㅋㅋㅋㅋㅋ)

남기다보니 주절주절 말이 많아졌는데 그만큼 생각을 많이 하게되고 의견이 점점 궁금해지는 재밌는 코드였습니다! 
다음 과제에서 진행할 호아님의 코드도 슬쩍 구경해야겠네요!! ㅋㅋㅋㅋ

리뷰는 제 개인적인 의견이고 잘잘못을 따지고자 남긴건 아니니 혹시라도 제가 부족하게 표현했다면 죄송스럽고, 
시간되실때 호아님 의견도 많이 남겨주세요! 😉

hoa0217 committed:
docs: 게임 진행 순서 수정

- 같은 문자 2개 -> n개

hoa0217 committed:
refactor(Record): existAllGreen -> existAnswer 메서드명

- 뷰에 의존적인 메서드명 수정

hoa0217 committed:
refactor(Results): isAllGreen -> isAnswer 메서드명

- 뷰에 의존적인 메서드명 수정

hoa0217 committed:
refactor(FileWordBook): find 메서드 반환 Optional로 수정

hoa0217 committed:
refactor(Alphabet): 알파벳 예외 메시지 추가

hoa0217 committed:
refactor(AnswerFormula): 정답 공식 예외 메시지 추가

hoa0217 committed:
refactor(FileReader): 파일 읽기 실패 예외 메시지 추가

hoa0217 committed:
refactor(Position): 올바르지 않은 위치 예외 메시지 추가

hoa0217 committed:
refactor(Wordle): 단어장에 단어가 없을 시 예외를 Wordle에서 던진다.

hoa0217 committed:
refactor(OutputView): 예외메세지를 받아서 뷰로 보여준다.

hoa0217 committed:
refactor(Wordle): 메서드 리팩토링
- processTurn -> processRound으로 변경
- isEnd대신 명시적으로 round가 끝났니?를 while 조건문에 걸어준다.
- while 중간에 정답을 맞추는 게임 성공 표시로 true를 리턴하여 성공종료 메시지를 콘솔에 보여준다.
- while 문이 끝날때까지 정답을 맞추지 못하면 false를 리턴하여 실패종료 메시지를 콘솔에 보여준다.

hoa0217 committed:
refactor(Record): 안쓰는 메서드 isEnd 삭제

hoa0217 committed:
refactor(Word): word 파라미터명 변경

- Word 클래스에서는 input과 answer를 구분하지 않는다. 따라서 파라미터명을 헷갈리지 않도록 변경한다.
- input -> word
- inputWord -> targetWord

hoa0217 committed:
docs: 용어 사전 및 클래스 다이어그램 수정

hoa0217 commented:
### 변경사항
- [x] `existAllGreen` -> `existAnswer`, `isAllGreen` -> `isAnswer` 메서드명 변경
  - 뷰에 의존적인 메서드명 변경
- [x] FileWordBook find 메서드 Optional 반환으로 수정
  - 어플리케이션단에서 예외처리 하도록 변경
- [x] Exception 메시지 추가하여 OutputConsole에 출력하도록 변경
- [x] Wordle runGame 메서드 리팩토링
  - 게임종료 체크메서드 분리 및 분기 처리 로직 추가
- [x] Word클래스 파라미터명 변경 (input -> target)  
- [x] 모델링 문서 수정
